### PR TITLE
Add EnterpriseUserAgent subspec with support for ExternalAgentCustomBrowser

### PR DESCRIPTION
**Pull request by soule**

### CODE REVIEW

**Comment by soule** [comment]
@yalkhder @WilliamDenniss Some of these changes are duplicated from the other PR (#540 ), and I would try to rebase off that or something before making the final PR.

However, how does this approach for adding EnterpriseUserAgent look? Thanks

**Comment by WilliamDenniss** [comment]
Looks good so far.

Does the subspec needs it's own subspec include header? In this case it would only include one other header, but conceivably there could be more enterprise user agents in the future.

**Comment by yalkhder** [comment]
> Looks good so far.
> 
> Does the subspec needs it's own subspec include header? In this case it would only include one other header, but conceivably there could be more enterprise user agents in the future.

Is there a realistic chance that more user agents will be created in the future? Also, will clients use multiple enterprise agents at the same time?

I'm trying to decide if a separate header for enterprise user agents will add more complexity without much benefit.

**Comment by WilliamDenniss** [comment]
Can you rebase on HEAD now that the other PR is merged?

File: Info.plist [hunk 1]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (yalkhder) [inline Info.plist:1]: Is this file needed? Why are we adding it now?

File: AppAuth.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -13,6 +13,71 @@
...
+		2D91B85A249053190005B197 /* OIDAuthorizationRequest.h in Headers */ = {isa = PBXBuildFile; fileRef = 341741B41C5D8243000EF209 /* OIDAuthorizationRequest.h */; settings = {ATTRIBUTES = (Public, ); }; };
+		2D91B85B249053190005B197 /* OIDTokenUtilities.h in Headers */ = {isa = PBXBuildFile; fileRef = 341741D51C5D8243000EF209 /* OIDTokenUtilities.h */; settings = {ATTRIBUTES = (Public, ); }; };
+		2D91B85C249053190005B197 /* OIDError.h in Headers */ = {isa = PBXBuildFile; fileRef = 341741BF1C5D8243000EF209 /* OIDError.h */; settings = {ATTRIBUTES = (Public, ); }; };
+		2D91B85D249053190005B197 /* OIDExternalUserAgentIOS.h in Headers */ = {isa = PBXBuildFile; fileRef = A6DEABA92018E5B50022AC32 /* OIDExternalUserAgentIOS.h */; settings = {ATTRIBUTES = (Public, ); }; };
+<<<<<<< HEAD
+		2D91B86B249177250005B197 /* AppAuthEnterpriseUserAgent.h in Headers */ = {isa = PBXBuildFile; fileRef = 2D91B86A249177250005B197 /* AppAuthEnterpriseUserAgent.h */; };
+=======
+>>>>>>> 218a2f1... Create dynamic targets
```
Comment (yalkhder) [inline AppAuth.xcodeproj/project.pbxproj:80]: Remove the conflict text (keep the actual line with AppAuthEnterpriseUserAgent.h).

File: AppAuth.podspec [hunk 3]
```diff
@@ -41,15 +41,15 @@ It follows the OAuth 2.0 for Native Apps best current practice
...
-     core.exclude_files = "Source/AppAuth.h"
+     core.source_files = "Source/AppAuthCore.h", "Source/AppAuthCore/*.{h,m}"
   end
 
   # Subspec for the full AppAuth library, including platform-dependant external user agents.
   s.subspec 'ExternalUserAgent' do |externalUserAgent|
     externalUserAgent.dependency 'AppAuth/Core'
     
     externalUserAgent.source_files = "Source/AppAuth.h", "Source/AppAuth/*.{h,m}"
+    externalUserAgent.exclude_files = "Source/AppAuth/iOS/OIDExternalUserAgentIOSCustomBrowser.{h,m}"
```
Comment (WilliamDenniss) [inline AppAuth.podspec:52]: I think it would be best if each sub-module was a separate folder.

Comment (WilliamDenniss) [reply AppAuth.podspec:52]: Awesome! Can remove this line now I think.

File: AppAuth.podspec [hunk 4]
```diff
@@ -62,5 +61,12 @@ It follows the OAuth 2.0 for Native Apps best current practice
     externalUserAgent.osx.deployment_target = '10.9'
   end
   
-  s.default_subspec = 'Core', 'ExternalUserAgent'
+  s.subspec 'EnterpriseUserAgent' do |enterpriseUserAgent|
+    enterpriseUserAgent.dependency 'AppAuth/Core'
+    
+    enterpriseUserAgent.source_files = "Source/AppAuthEnterpriseUserAgent.h", "Source/AppAuthEnterpriseUserAgent/iOS/OIDExternalUserAgentIOSCustomBrowser.{h,m}"
```
Comment (WilliamDenniss) [inline AppAuth.podspec:67]: Should it be this?
` enterpriseUserAgent.ios.source_files = "Source/AppAuthEnterpriseUserAgent/iOS/**/*.{h,m}"`

File: Source/EnterpriseUserAgentFramework/AppAuthEnterpriseUserAgent.h [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+       distributed under the License is distributed on an "AS IS" BASIS,
+       WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+       See the License for the specific language governing permissions and
+       limitations under the License.
+*/
+
+#import <Foundation/Foundation.h>
+
+//! Project version number for AppAuthFramework-iOS.
+FOUNDATION_EXPORT double AppAuthCoreVersionNumber;
```
Comment (WilliamDenniss) [inline Source/EnterpriseUserAgentFramework/AppAuthEnterpriseUserAgent.h:22]: Should this be exporting the AppAuthCoreVersionNumber? Or AppAuthEnterpriseUserAgentVersionNumber?

**Comment by WilliamDenniss** [comment]
I think the README should be updated to note that these are now in a sub-spec (CocoaPod) and separate framework (Carthage) https://github.com/openid/AppAuth-iOS#custom-browser-user-agent

File: Source/EnterpriseUserAgentFramework/AppAuthEnterpriseUserAgent.h (hunk 5 above)
Comment (soule) [reply Source/EnterpriseUserAgentFramework/AppAuthEnterpriseUserAgent.h:22]: Good catch, thanks.

File: AppAuth.podspec (hunk 4 above)
Comment (soule) [reply AppAuth.podspec:67]: Good point, that will allow for future extensibility. I think we should still include the header file explicitly

File: Source/EnterpriseUserAgentFramework/AppAuthEnterpriseUserAgent.h [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+#import <AppAuthEnterpriseUserAgent/OIDScopes.h>
+#import <AppAuthEnterpriseUserAgent/OIDScopeUtilities.h>
+#import <AppAuthEnterpriseUserAgent/OIDServiceConfiguration.h>
+#import <AppAuthEnterpriseUserAgent/OIDServiceDiscovery.h>
+#import <AppAuthEnterpriseUserAgent/OIDTokenRequest.h>
+#import <AppAuthEnterpriseUserAgent/OIDTokenResponse.h>
+#import <AppAuthEnterpriseUserAgent/OIDTokenUtilities.h>
+#import <AppAuthEnterpriseUserAgent/OIDURLSessionProvider.h>
+#import <AppAuthEnterpriseUserAgent/OIDEndSessionRequest.h>
+#import <AppAuthEnterpriseUserAgent/OIDEndSessionResponse.h>
```
Comment (WilliamDenniss) [inline Source/EnterpriseUserAgentFramework/AppAuthEnterpriseUserAgent.h:52]: What's the rationale for the EnterpriseUserAgentFramework including all the AppAuth headers?

File: README.md [hunk 7]
```diff
@@ -506,6 +506,22 @@ try, or use as a reference for your own implementation. One of them,
...
+##### CocoaPods
+Include the `EnterpriseUserAgent` subspec alongside any pods/subspecs you were already using, e.g.:
+```
+	pod 'AppAuth'
+	pod 'AppAuth/EnterpriseUserAgent'
+```
+Make sure to import `AppAuthEnterpriseUserAgent` in addition to `AppAuth` if you are using the full `AppAuth` functionality.
+
+##### Carthage
+Use the `EnterpriseUserAgent` Framework, which includes all the headers of the `AppAuth` framework.
```
Comment (WilliamDenniss) [inline README.md:518]: I think we should call the framework 'AppAuthEnterpriseUserAgent' to match CocoaPods documented above, and SPM documented below.

File: Source/EnterpriseUserAgentFramework/AppAuthEnterpriseUserAgent.h (hunk 6 above)
Comment (WilliamDenniss) [reply Source/EnterpriseUserAgentFramework/AppAuthEnterpriseUserAgent.h:52]: Resolved offline, each of AppAuth and AppAuthEnterpriseUserAgent are the fully library (and mutually exclusive)

**Comment by WilliamDenniss** [comment]
LGTM
